### PR TITLE
#1398 - treat receive buffer smaller than MTU as zero window

### PR DESCRIPTION
**Pull request by majek**
When receiving data, netstack avoids sending spurious acks. When
user does recv() should netstack send ack telling the sender that
the window was increased? It depends. Before this patch, netstack
_will_ send the ack in the case when window was zero or window >>
scale was zero. Basically - when recv space increased from zero.

This is not working right with silly-window-avoidance on the sender
side. Some network stacks refuse to transmit segments, that will fill
the window but are below MSS. Before this patch, this confuses
netstack. On one hand if the window was like 3 bytes, netstack
will _not_ send ack if the window increases. On the other hand
sending party will refuse to transmit 3-byte packet.

This patch changes that, making netstack treat a window below MTU
as a zero window.

Signed-off-by: Marek Majkowski <marek@cloudflare.com>

### CODE REVIEW

**Comment by majek** [comment]
Fixes #1398

**Comment by majek** [comment]
@hbhasker please review again. Changes:
 - used `amss` as suggested.
 - guarded the sent ack to only when the receive buffer grows to >= amss (this should prevent silly window)
 - introduced TestIncreaseWindowOnReceive test, which should be rather straightforward. This is a modifed "TestZeroScaledWindowReceive test. Surprisingly, it tests almost the same thing, with the small difference - zeroscaledwindow actually forces the avail buffer to be zero. We don't. (just < 1MSS)

File: pkg/tcpip/transport/tcp/tcp_test.go [hunk 1]
```diff
@@ -6561,3 +6561,71 @@ func TestKeepaliveWithUserTimeout(t *testing.T) {
...
+
+func TestIncreaseWindowOnReceive(t *testing.T) {
+	// This test ensures that the endpoint sends an ack,
+	// after recv() when the window grows to more than 1 MSS.
+	c := context.New(t, defaultMTU)
+	defer c.Cleanup()
+
+	// Set the window size such that a window scale of 4 will be
+	// used. Testing window scaling is outside scope of this test.
+	const wnd = 65535 * 10
```
Comment (hbhasker) [inline pkg/tcpip/transport/tcp/tcp_test.go]: rename wnd w/ rcvBuf? Since we are setting the receiver buffer with this not the window itself.

File: pkg/tcpip/transport/tcp/endpoint.go [hunk 2]
```diff
@@ -1138,11 +1148,8 @@ func (e *endpoint) Peek(vec [][]byte) (int64, tcpip.ControlMessages, *tcpip.Erro
 //
 // It must be called with rcvListMu held.
 func (e *endpoint) zeroReceiveWindow(scale uint8) bool {
```
Comment (hbhasker) [inline pkg/tcpip/transport/tcp/endpoint.go]: Shouldn't this also account for cases where scaling is disabled? Don't we want this to return true if the available space goes below that of MSS?

File: pkg/tcpip/transport/tcp/tcp_test.go [hunk 3]
```diff
@@ -6561,3 +6561,71 @@ func TestKeepaliveWithUserTimeout(t *testing.T) {
 		t.Errorf("got c.Stack().Stats().TCP.EstablishedTimedout = %v, want = %v", got, want)
 	}
 }
+
+func TestIncreaseWindowOnReceive(t *testing.T) {
+	// This test ensures that the endpoint sends an ack,
+	// after recv() when the window grows to more than 1 MSS.
```
Comment (hbhasker) [inline pkg/tcpip/transport/tcp/tcp_test.go]: Given that the original bug was when window scaling was disabled we should probably have a subtest or a separate test that tests the correct behaviour when window scaling is disabled.

Another edge case that I feel we need to verify is what happens if the user sets the rcvBuf to a value far lower than MSS. eg

Set the rcvBuf to like 4KB when MSS is 65kb.

File: pkg/tcpip/transport/tcp/tcp_test.go (hunk 1 above)
Comment (majek) [reply pkg/tcpip/transport/tcp/tcp_test.go]: done

File: pkg/tcpip/transport/tcp/endpoint.go (hunk 2 above)
Comment (majek) [reply pkg/tcpip/transport/tcp/endpoint.go]: window scaling: I removed it from test, it's irrelvant here.

Negative. We keep the semantics of zero-window exactly as they were. Instead we redefine the semantics of what should happen on recv() when window is small but non-zero.

In past: recv() would trigger ACK _only_ when window was in zero-window mode.
Now: recv() will do that, but also trigger ACK when window grew to >= 1MSS.

The difference is subtle, and probably wrong - although it's another discussion if we want to fix this. Consider recv() returning small value in three cases:
 - (A) recv()=small when window is zero
 - (B) recv()=small when window is one
 - (C) recv()=mss when window is one

Before this patch (A) would generate ACK, (B),(C) would not. With this patch (A) and (C) genrate ack, (B) does not.

Arguably, we could implement fully recv side silly window syndrome avoidance and prevent (A) from generating ACK if the avail is small.... not sure if this is part of this bug/stream of work.

File: pkg/tcpip/transport/tcp/tcp_test.go (hunk 3 above)
Comment (majek) [reply pkg/tcpip/transport/tcp/tcp_test.go]: I removed window scaling from the test, it's really irrelevant.

If user sets recv buffer < MSS, it's her problem. This code won't kick in. This code only makes sure that if buffer (avail) was small, and we did recv() and it's now large, we should immediately ack it, to allow sender to uncork itself (and help senders side silly-window prevention).

I added a code and test to make sure that _increasing_ the recv buffer can indeed generate ack as it should if we suddenly have > MSS avail.

File: pkg/tcpip/transport/tcp/endpoint.go (hunk 2 above)
Comment (hbhasker) [reply pkg/tcpip/transport/tcp/endpoint.go]: Won't the correct behaviour be for A and B to not generate a window update since there isn't enough space available? I believe only C) should generate a window update right to avoid silly-window. Eventually a future read should clear enough space and cause a zero window to be generated due to > MSS space being available right?

Comment (majek) [reply pkg/tcpip/transport/tcp/endpoint.go]: Correct, but I'm not totally sure we want to address (A) in this PR. Handling (A) might require some more guarantees (like ensuring recvbufSize > MSS).
Also, current semantics are not wrong, they may be just considered wasteful (spurious acks). For (C) which this PR addresses the semantics as they stand now are plain wrong.

Comment (hbhasker) [reply pkg/tcpip/transport/tcp/endpoint.go]: Agreed. I am fine with this change as-is and we can address A separately.
